client/stats/host.go

```diff
@@ -0,0 +1,162 @@
+package stats
+
+import (
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/disk"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/mem"
+)
+
+// HostStats represents resource usage stats of the host running a Nomad client
+type HostStats struct {
+	Memory    *MemoryStats
+	CPU       []*CPUStats
+	DiskStats []*DiskStats
+	Uptime    uint64
+}
+
+// MemoryStats represnts stats related to virtual memory usage
+type MemoryStats struct {
+	Total     uint64
+	Available uint64
+	Used      uint64
+	Free      uint64
+}
+
+// CPUStats represents stats related to cpu usage
+type CPUStats struct {
+	CPU    string
+	User   float64
+	System float64
+	Idle   float64
+	Total  float64
+}
+
+// DiskStats represents stats related to disk usage
+type DiskStats struct {
+	Device            string
+	Mountpoint        string
+	Size              uint64
+	Used              uint64
+	Available         uint64
+	UsedPercent       float64
+	InodesUsedPercent float64
+}
+
+// HostStatsCollector collects host resource usage stats
+type HostStatsCollector struct {
+	statsCalculator map[string]*HostCpuStatsCalculator
+}
+
+// NewHostStatsCollector returns a HostStatsCollector
+func NewHostStatsCollector() (*HostStatsCollector, error) {
+	times, err := cpu.Times(true)
+	if err != nil {
+		return nil, err
+	}
+	statsCalculator := make(map[string]*HostCpuStatsCalculator)
+	for _, time := range times {
+		statsCalculator[time.CPU] = NewHostCpuStatsCalculator()
+	}
+	return &HostStatsCollector{statsCalculator: statsCalculator}, nil
+}
+
+// Collect collects stats related to resource usage of a host
+func (h *HostStatsCollector) Collect() (*HostStats, error) {
+	hs := &HostStats{}
+	if memStats, err := mem.VirtualMemory(); err == nil {
+		ms := &MemoryStats{
+			Total:     memStats.Total,
+			Available: memStats.Available,
+			Used:      memStats.Used,
+			Free:      memStats.Free,
+		}
+		hs.Memory = ms
+	}
+
+	if cpuStats, err := cpu.Times(true); err == nil {
+		cs := make([]*CPUStats, len(cpuStats))
+		for idx, cpuStat := range cpuStats {
+			cs[idx] = &CPUStats{
+				CPU:    cpuStat.CPU,
+				User:   cpuStat.User,
+				System: cpuStat.System,
+				Idle:   cpuStat.Idle,
+			}
+			if percentCalculator, ok := h.statsCalculator[cpuStat.CPU]; ok {
+				idle, user, system, total := percentCalculator.Calculate(cpuStat)
+				cs[idx].Idle = idle
+				cs[idx].System = system
+				cs[idx].User = user
+				cs[idx].Total = total
+			} else {
+				h.statsCalculator[cpuStat.CPU] = NewHostCpuStatsCalculator()
```

Comment:
Is this guarding the  same thing as NewHostStatsCollector is doing. If so why not just do it lazily here